Assets/Documentation/man/man3/SpawnPiano.3
.TH "SpawnPiano" 3 "Version 1.0.0" "KiBoard GDD & Technical Documentation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SpawnPiano
.SH SYNOPSIS
.br
.PP
.PP
Inherits MonoBehaviour\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "int \fBGetIndexOfSpriteRenderer\fP (SpriteRenderer sr)"
.br
.RI "gets the index of a specific keys sprite renderer\&. "
.ti -1c
.RI "Color \fBGetDefaultKeyColour\fP (int i)"
.br
.RI "gets default colour for a key based on key number\&. "
.ti -1c
.RI "void \fBClearAllKeyColours\fP ()"
.br
.RI "resets colours of all keys on the keyboard\&. "
.ti -1c
.RI "void \fBUpdateKeyColours\fP (int noteNumber, bool enabled, string timingScore='')"
.br
.RI "Updates the colours of keys when notes are pressed based off of the timing score\&. "
.ti -1c
.RI "void \fBSpawnKeyParticle\fP (int keyNum, string score='')"
.br
.RI "spawn particle effect at pressed key\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static Color \fBGetKeyColour\fP (int keyNum, string timingScore='')"
.br
.RI "Get colour of a key based off of timing score\&. "
.ti -1c
.RI "static Color \fBGetKeyColour\fP (int keyNum, bool isKeyEnabled, string timingScore='')"
.br
.RI "Get colour of a key based off of timing score taking in to account notes being released\&. "
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static \fBSpawnPiano\fP \fBinstance\fP"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBAwake\fP ()"
.br
.in -1c
.SS "Static Private Member Functions"

.in +1c
.ti -1c
.RI "static bool \fBCheckBlackNote\fP (int i)"
.br
.RI "is the note at index I a black note? "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "GameObject \fBpianoWhiteTilePrefab\fP"
.br
.ti -1c
.RI "GameObject \fBkeyParticlePrefab\fP"
.br
.ti -1c
.RI "SpriteRenderer[] \fBspriterenderers\fP = new SpriteRenderer[88]"
.br
.RI "array of sprites for each key of the piano\&. "
.ti -1c
.RI "Color \fBlane1\fP"
.br
.RI "Lane colour 1\&. "
.ti -1c
.RI "Color \fBlane2\fP"
.br
.RI "lane colour 2\&. "
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static Color \fBenabledColour\fP = new Color(255, 0, 0, 128)"
.br
.RI "colour when incorrectly pressed/timed\&. "
.ti -1c
.RI "static Color \fBperfectColour\fP = new Color(255, 0, 195, 128)"
.br
.RI "colour when perfectly timed "
.ti -1c
.RI "static Color \fBgoodColour\fP = new Color(0\&.15f, 1, \&.5f, 128)"
.br
.RI "colour when pressed with good timing\&. "
.ti -1c
.RI "static Color \fBokayColour\fP = new Color(0, 9, 255, 128)"
.br
.RI "colour when pressed with okay timing\&. "
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "void SpawnPiano\&.Awake ()\fR [private]\fP"

.nf
42     {
43         if (instance == null) { instance = this; }
44         else { Destroy(instance\&.gameObject); }
45     }
.PP
.fi

.SS "static bool SpawnPiano\&.CheckBlackNote (int i)\fR [static]\fP, \fR [private]\fP"

.PP
is the note at index I a black note? 
.nf
95     {
96         int value = i % 12;   
97         if(GameSettings\&.usePiano) return value == 2 || value == 5 || value == 7 || value == 10 || (value == 0 && i != 0);
98         return value == 0 || value == 2 ||value == 5 || value == 7 || value == 10 ;
99 
100     }
.PP
.fi

.PP
Referenced by \fBGetDefaultKeyColour()\fP, and \fBGetKeyColour()\fP\&.
.SS "void SpawnPiano\&.ClearAllKeyColours ()"

.PP
resets colours of all keys on the keyboard\&. 
.nf
85     {
86         for (int i = 0; i < 88; i++)
87         {
88             spriterenderers[i]\&.color = GetDefaultKeyColour(i);
89         }
90     }
.PP
.fi

.SS "Color SpawnPiano\&.GetDefaultKeyColour (int i)"

.PP
gets default colour for a key based on key number\&. 
.PP
\fBParameters\fP
.RS 4
\fIi\fP 
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.nf
78     {
79         return CheckBlackNote(i + 1) ? Color\&.black : Color\&.white;
80     }
.PP
.fi

.PP
Referenced by \fBClearAllKeyColours()\fP\&.
.SS "int SpawnPiano\&.GetIndexOfSpriteRenderer (SpriteRenderer sr)"

.PP
gets the index of a specific keys sprite renderer\&. 
.PP
\fBParameters\fP
.RS 4
\fIsr\fP 
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.nf
69     {
70         return Array\&.IndexOf(spriterenderers, sr);
71     }
.PP
.fi

.SS "static Color SpawnPiano\&.GetKeyColour (int keyNum, bool isKeyEnabled, string timingScore = \fR''\fP)\fR [static]\fP"

.PP
Get colour of a key based off of timing score taking in to account notes being released\&. 
.nf
145     {
146 
147         if (!isKeyEnabled)
148         {
149             if (CheckBlackNote(keyNum + 1)) { return Color\&.black; }
150             else { return Color\&.white; }
151         }
152         switch (timingScore)
153         {
154             case "Perfect":
155                 return perfectColour;
156 
157             case "Good":
158                 return goodColour;
159 
160             case "Okay":
161                 return okayColour;
162             default:
163                 return enabledColour;
164 
165         }
166 
167     }
.PP
.fi

.SS "static Color SpawnPiano\&.GetKeyColour (int keyNum, string timingScore = \fR''\fP)\fR [static]\fP"

.PP
Get colour of a key based off of timing score\&. 
.nf
122     {
123         switch (timingScore)
124         {
125             case "Perfect":
126                 return perfectColour;
127 
128             case "Good":
129                 return goodColour;
130 
131             case "Okay":
132                 return okayColour;
133 
134             default:
135                 return enabledColour;
136 
137         }
138 
139     }
.PP
.fi

.PP
Referenced by \fBSpawnKeyParticle()\fP, and \fBUpdateKeyColours()\fP\&.
.SS "void SpawnPiano\&.SpawnKeyParticle (int keyNum, string score = \fR''\fP)"

.PP
spawn particle effect at pressed key\&. 
.PP
\fBParameters\fP
.RS 4
\fIkeyNum\fP key number to spawn the particles at\&.
.br
\fIscore\fP timing score used for colouring the particles\&.
.RE
.PP

.nf
175     {
176         ParticleSystem particle;
177         if (GameSettings\&.usePiano) { particle = Instantiate(keyParticlePrefab, spriterenderers[keyNum]\&.transform)\&.GetComponent<ParticleSystem>(); }
178         else { particle = Instantiate(keyParticlePrefab, spriterenderers[(keyNum\-3 )%12]\&.transform)\&.GetComponent<ParticleSystem>(); }
179         particle\&.transform\&.position += Vector3\&.up / 2;
180         ParticleSystem\&.MainModule main = particle\&.main;
181         main\&.startColor = GetKeyColour(keyNum, score);
182         particle\&.Play();
183         Destroy(particle, 5f);
184     }
.PP
.fi

.SS "void SpawnPiano\&.UpdateKeyColours (int noteNumber, bool enabled, string timingScore = \fR''\fP)"

.PP
Updates the colours of keys when notes are pressed based off of the timing score\&. 
.PP
\fBParameters\fP
.RS 4
\fInoteNumber\fP note number of the note pressed\&.
.br
\fIenabled\fP is the note being enabled or disabled?
.br
\fItimingScore\fP timing score of the note\&.
.RE
.PP

.nf
108     {
109         if (GameSettings\&.usePiano)
110         {
111             spriterenderers[noteNumber]\&.color = GetKeyColour(noteNumber, enabled, timingScore);
112         }
113         else
114         {
115             spriterenderers[(noteNumber\-3 )% 12]\&.color = GetKeyColour(noteNumber, enabled, timingScore);
116         }
117     }
.PP
.fi

.SH "Member Data Documentation"
.PP 
.SS "Color SpawnPiano\&.enabledColour = new Color(255, 0, 0, 128)\fR [static]\fP, \fR [private]\fP"

.PP
colour when incorrectly pressed/timed\&. 
.PP
Referenced by \fBGetKeyColour()\fP, and \fBGetKeyColour()\fP\&.
.SS "Color SpawnPiano\&.goodColour = new Color(0\&.15f, 1, \&.5f, 128)\fR [static]\fP, \fR [private]\fP"

.PP
colour when pressed with good timing\&. 
.PP
Referenced by \fBGetKeyColour()\fP, and \fBGetKeyColour()\fP\&.
.SS "\fBSpawnPiano\fP SpawnPiano\&.instance\fR [static]\fP"

.PP
Referenced by \fBSongNoteEditor\&.AddNote()\fP, \fBAwake()\fP, \fBMidiInput\&.NoteOff()\fP, \fBMidiInput\&.NoteOn()\fP, \fBMidiInput\&.OnNoteSuccess()\fP, and \fBMidiInput\&.PedalStateChanged()\fP\&.
.SS "GameObject SpawnPiano\&.keyParticlePrefab\fR [private]\fP"

.PP
Referenced by \fBSpawnKeyParticle()\fP\&.
.SS "Color SpawnPiano\&.lane1\fR [private]\fP"

.PP
Lane colour 1\&. 
.SS "Color SpawnPiano\&.lane2\fR [private]\fP"

.PP
lane colour 2\&. 
.SS "Color SpawnPiano\&.okayColour = new Color(0, 9, 255, 128)\fR [static]\fP, \fR [private]\fP"

.PP
colour when pressed with okay timing\&. 
.PP
Referenced by \fBGetKeyColour()\fP, and \fBGetKeyColour()\fP\&.
.SS "Color SpawnPiano\&.perfectColour = new Color(255, 0, 195, 128)\fR [static]\fP, \fR [private]\fP"

.PP
colour when perfectly timed 
.PP
Referenced by \fBGetKeyColour()\fP, and \fBGetKeyColour()\fP\&.
.SS "GameObject SpawnPiano\&.pianoWhiteTilePrefab\fR [private]\fP"

.SS "SpriteRenderer [] SpawnPiano\&.spriterenderers = new SpriteRenderer[88]\fR [private]\fP"

.PP
array of sprites for each key of the piano\&. 
.PP
Referenced by \fBClearAllKeyColours()\fP, \fBGetIndexOfSpriteRenderer()\fP, \fBSpawnKeyParticle()\fP, and \fBUpdateKeyColours()\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for KiBoard GDD & Technical Documentation from the source code\&.


Assets/Scripts/SpawnPiano.cs
using JetBrains.Annotations;
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class SpawnPiano : MonoBehaviour
{
    public static SpawnPiano instance;
    [SerializeField] GameObject pianoWhiteTilePrefab;
    [SerializeField] GameObject keyParticlePrefab;
    /// <summary>
    /// array of sprites for each key of the piano.
    /// </summary>
    [SerializeField] SpriteRenderer[] spriterenderers = new SpriteRenderer[88];
    /// <summary>
    /// colour when incorrectly pressed/timed.
    /// </summary>
    [SerializeField] static Color enabledColour = new Color(1f, 0, 0, 1);
    [SerializeField] static Color enabledColourBlack = new Color(.8f, 0, 0, 1);
    /// <summary>
    /// colour when perfectly timed
    /// </summary>
    [SerializeField] static Color perfectColour = new Color(1, 0, .76f, 1);
    /// <summary>
    /// colour when pressed with good timing.
    /// </summary>
    [SerializeField] static Color goodColour = new Color(0.15f, 1, .5f, 1);
    /// <summary>
    /// colour when pressed with okay timing.
    /// </summary>
    [SerializeField] static Color okayColour = new Color(0, .04f, 1, 1);
    /// <summary>
    /// Lane colour 1.
    /// </summary>
    [SerializeField] Color lane1;
    /// <summary>
    /// lane colour 2.
    /// </summary>
    [SerializeField] Color lane2;

    private void Awake()
    {
        if (instance == null) { instance = this; }
        else { Destroy(instance.gameObject); }
    }
    /* unused, used to spawn piano tiles and lanes during runtime.
        void spawnPiano()
        {
            for (int i = 0; i < 88; i++)
            {
                // create piano tile, set position, and colour every other one.
                Transform tile = Instantiate(pianoWhiteTilePrefab, transform.position + Vector3.right * 0.2045455f * (i), Quaternion.identity).transform;
                tile.SetParent(transform.parent);
                spriterenderers[i] = tile.GetComponent<SpriteRenderer>();
                spriterenderers[i].color = GetDefaultKeyColour(i);
                var lane = tile.GetChild(0).GetComponent<SpriteRenderer>();
                lane.color = i % 2 == 0 ? lane1 : lane2;
                lane.transform.position += (Vector3.up * lane.transform.localScale.y/2);


            }
        }*/
    /// <summary>
    /// gets the index of a specific keys sprite renderer.
    /// </summary>
    /// <param name="sr"></param>
    /// <returns></returns>
    public int GetIndexOfSpriteRenderer(SpriteRenderer sr)
    {
        return Array.IndexOf(spriterenderers, sr);
    }
    /// <summary>
    /// gets default colour for a key based on key number.
    /// </summary>
    /// <param name="i"></param>
    /// <returns></returns>
    public Color GetDefaultKeyColour(int i)
    {
        return CheckBlackNote(i + 1) ? Color.black : Color.white;
    }
    /// <summary>
    /// resets colours of all keys on the keyboard.
    /// </summary>
    public void ClearAllKeyColours()
    {
        for (int i = 0; i < 88; i++)
        {
            spriterenderers[i].color = GetDefaultKeyColour(i);
        }
    }
    /// <summary>
    /// is the note at index I a black note?
    /// </summary>
    static bool CheckBlackNote(int i)
    {
        int value = i % 12;   
        if(GameSettings.usePiano) return value == 2 || value == 5 || value == 7 || value == 10 || (value == 0 && i != 0);
        return value == 0 || value == 2 ||value == 5 || value == 7 || value == 10 ;

    }
    /// <summary>
    /// Updates the colours of keys when notes are pressed based off of the timing score.
    /// </summary>
    /// <param name="noteNumber"> note number of the note pressed.</param>
    /// <param name="enabled"> is the note being enabled or disabled?</param>
    /// <param name="timingScore"> timing score of the note.</param>
    public void UpdateKeyColours(int noteNumber, bool enabled, string timingScore = "")
    {
        if (GameSettings.usePiano)
        {
            spriterenderers[noteNumber].color = GetKeyColour(noteNumber, enabled, timingScore);
        }
        else
        {
            spriterenderers[(noteNumber-3 )% 12].color = GetKeyColour(noteNumber, enabled, timingScore);
        }
    }
    /// <summary>
    /// Get colour of a key based off of timing score.
    /// </summary>
    public static Color GetKeyColour(int keyNum, string timingScore = "")
    {
        switch (timingScore)
        {
            case "Perfect":
                return perfectColour;

            case "Good":
                return goodColour;

            case "Okay":
                return okayColour;

            default:
                if (CheckBlackNote(keyNum + 1)) { return enabledColourBlack; }
                return enabledColour;

        }

    }
    /// <summary>
    /// Get colour of a key based off of timing score taking in to account notes being released.
    /// </summary>

    public static Color GetKeyColour(int keyNum, bool isKeyEnabled, string timingScore = "")
    {

        if (!isKeyEnabled)
        {
            if (CheckBlackNote(keyNum + 1)) { return Color.black; }
            else { return Color.white; }
        }
        switch (timingScore)
        {
            case "Perfect":
                return perfectColour;

            case "Good":
                return goodColour;

            case "Okay":
                return okayColour;
            default:
                if (CheckBlackNote(keyNum + 1)) { return enabledColourBlack; }
                return enabledColour;

        }

    }

    /// <summary>
    /// spawn particle effect at pressed key.
    /// </summary>
    /// <param name="keyNum">key number to spawn the particles at.</param>
    /// <param name="score"> timing score used for colouring the particles.</param>
    public void SpawnKeyParticle(int keyNum, string score = "")
    {
        ParticleSystem particle;
        if (GameSettings.usePiano) { particle = Instantiate(keyParticlePrefab, spriterenderers[keyNum].transform).GetComponent<ParticleSystem>(); }
        else { particle = Instantiate(keyParticlePrefab, spriterenderers[(keyNum-3 )%12].transform).GetComponent<ParticleSystem>(); }
        particle.transform.position += Vector3.up / 2;
        ParticleSystem.MainModule main = particle.main;
        main.startColor = GetKeyColour(keyNum, score);
        particle.Play();
        Destroy(particle, 5f);
    }
}
